### PR TITLE
Create CODEOWNERS

### PR DESCRIPTION
**Pull request by Mariatta**
Copied over info from .mention-bot

https://github.com/python/core-workflow/issues/160

### CODE REVIEW

**Comment by Mariatta** [comment]
:thinking: wondering if lowercase `codeowners` file will work ...

File: .github/CODEOWNERS
Comment (brettcannon) [inline .github/CODEOWNERS]: Could this be collapsed with `Lib/hashlib.py` with `hashlib.*`? Same for `ssl.*`? And if you really want to catch everything including tests with one rule then would `*hashlib.*` work?

File: .github/CODEOWNERS
Comment (brettcannon) [inline .github/CODEOWNERS]: `pyhash.*` should cover both the `.c` and `.h` files.

File: .github/CODEOWNERS
Comment (brettcannon) [inline .github/CODEOWNERS]: Since this file might grow a lot, we should probably have a comment per section to say what the rules are meant for, e.g. `# Hashing`.

**Review by brettcannon — COMMENTED** [review]
Might as well delete `.mention-bot` while we're at it.

**Comment by brettcannon** [comment]
@tiran since he would be the inaugural user of the file. 😄

**Comment by Mariatta** [comment]
mention-bot deletion is in https://github.com/python/cpython/pull/2923 :)

**Comment by brettcannon** [comment]
Then if @tiran is happy with the results and we add a section comment so we know thematically what the rules are meant to be for, then it LGTM to merge this and delete the `.mention-bot` file!

**Comment by tiran** [comment]
Could we auto-generate CODEOWNERS from .mentionbot? Or have a script that generates both from a simple rule file? Also mentionbot configuration and code ownership are not quite the same. For example I'm just interested to get notified about pyhash changes. The actual owners are @haypo and @pitrou .

**Comment by pitrou** [comment]
I really dislike the "owners" terminology as used here. There are "experts", which is the terminology used in the devguide. I personally don't want some people to start being possessive towards code they contributed (any more than they might already be).

**Review by tiran — APPROVED** [review]
I'm happy to be the guinea pig for your experiments. :)

**Comment by tiran** [comment]
Instead of assigning individual users to files, could we create teams and assign teams to files instead?

@pitrou yes, I agree with you. The term "owner" is problematic. However we can't do much about it. It's a github standard, https://help.github.com/articles/about-codeowners/

**Comment by Mariatta** [comment]
Reading codeowner docs, it `@org/team-name` is supported, so we can already do:
`@python/python-core`, `@python/asyncio`, `@python/typing`.
If we go this path, we may need to create many new teams, and people will need to request to be added to/removed from tho team.
I think it's simpler to let individuals add/remove their usernames in `codeowners`, they can create the PR anytime.

**Comment by tiran** [comment]
@Mariatta the ``CODEOWNERS`` file may become hard to maintain pretty quickly. The file format has a last-match-wins ruling. Without teams it's going to be hard to maintain ownership for e.g. ``Docs/*`` for some and specific docs for others.

File: .github/CODEOWNERS
Comment (Mariatta) [reply .github/CODEOWNERS]: So I'm thinking to replace with the following patterns. What do you think? cc @tiran 
```
# Hashing
**/*hashlib*                  @tiran
**/*pyhash*                   @tiran

# SSL
**/*ssl*                      @tiran

```

**Comment by brettcannon** [comment]
@pitrou Yes, the terminology doesn't line up with what we will use the file for, but it's dictated by GitHub (they actually have an option to say a pull request can't be merged until an owner listed in that file has been cleared, hence the naming). We could have a comment at the top of the file to point this fact out if you want.

@tiran If we use a `CODEOWNERS` file then we're ditching mention-bot since the latter isn't working right now (not sure if you've noticed but it's actually been broken for well over a month now). Basically if we can rely on something that GitHub provides as a service rather than something Facebook provides just to be nice then I'll go with the one that has paid staffing to make sure it works. 😄 Otherwise we will need to stand up our own mention-bot instance and maintain it.

As for having teams, my plan was to get this file in first, and then solicit people to tell me what teams they wanted created as part of the announcement that the file now existed. I would then create the team and then make all team members admins for that team so they can control when they leave, adding others, etc. That means I only have to create the initial team and past that I don't have to maintain it. But one thing at a time. 😄 

And I hope no one is silly enough to add a `Docs/*` rule as that would be extremely noisy. 😉

**Comment by Mariatta** [comment]
I pushed the change. Please re-review.

**Review by vstinner — CHANGES_REQUESTED** [review]
I don't understand the purpose of this file. Please write a short description maybe pointing to a full description somewhere else.

Does it technically prevent merging any kind of change in these files without @tiran approval?

I see a risk of abusing this file.

**Comment by vstinner** [comment]
https://help.github.com/articles/about-codeowners/ mentions adding automatically requests reviews. That's fine.

My question is about the optional part: do you want to require @tiran review? Or later require someone else review on other parts of the code.

**Comment by Mariatta** [comment]
In the current `.mention-bot` file, it has @tiran in "AlwaysNotifyForPaths"  for these paths.
So I'm adding him here to `CODEOWNERS` for backward compatibility 😄

**Comment by Mariatta** [comment]
I added some comments on top of the file.

**Comment by Mariatta** [comment]
I think for now we are only notifying the owner/expert listed in `codeowners` file, but we are not requiring their review.

**Comment by Mariatta** [comment]
Not sure if people want to be requested for review for backport PRs? I removed `needs backport to <branch>` labels.

**Comment by Mariatta** [comment]
😞 I feel terrible that I made two personal branches in python/cpython repo that lasted longer than 24 hours. So I'm going to merge this now 😛 
And if the format/pattern in this file doesn't work out, please propose a new PR 😃 Thanks all!

**Comment by brettcannon** [comment]
To answer @haypo : it only blocks PRs if you check a certain box and I'm never going to check that box (trust me 😉 )

**Comment by vstinner** [comment]
Brett: "I'm never going to check that box (trust me 😉 )"

Oh ok, that's fine in this case ;-)

**Pull request merged**
